### ISSUE TITLE
.collapseToEnd() method of Transform not collapsing to end

### ISSUE THREAD

**Issue by laurenkt** [ianstormtaylor/slate#611]
(Browser tested on: Safari 10.0.3)

An example of this is on the [Links example](http://slatejs.org/#/links?_k=065jie).

When adding a link to selected text the transform has a `.collapseToEnd()`. My understanding is that this positions the selection point to the end of the selection, which seems to be what happens when you select a word and add a link.

However, when selecting an entire block, e.g. a paragraph on that example, and inserting a link, the selection instead gets positioned at the beginning of the selection not at the end, despite the call to `.collapseToEnd()`.

(Secondary to this, the cursor does not actually show at the location that the real selection is).

Here's a video of what I mean (simply pressing the 'a' button after a link has been made without moving the selection point): https://youtu.be/n3o2Is9u20M – first it shows the behaviour I expect to happen in all cases, then it shows what actually happens in the specific case when a whole block is selected.

**Comment by Slapbox** [comment]
Can't recreate in Chrome on Windows 7, but your video shows it is obviously real.

Any console errors from that?

**Comment by laurenkt** [comment]
No console errors !

**Comment by laurenkt** [comment]
Hmm so I just tried on Chrome and the problem does exist there. There is not also the additional problem of the cursor showing in a different place from where it actually is on Chrome. But there is still the issue of it jumping to the end of selection normally but to the _start_ for whole blocks.

**Comment by Slapbox** [comment]
Yes I see what you mean about it jumping to the beginning of blocks if you select the whole thing.

**Comment by skogsmaskin** [comment]
I wonder if this is related to that you cant select the last character of the link if the link is the last part of the string.

Try deleting the '!' from the link in the link example, and try to select the 's' backwards. In will not select in Chrome at least.

**Comment by oyeanuj** [comment]
Here is a simple GIF reproducing the problem on the examples: 
![slate-selection-error](https://cloud.githubusercontent.com/assets/9633371/23591977/985f213a-01ae-11e7-8623-c7c07f54405e.gif)

#### Steps:
1. Select some text in a paragraph with inline nodes.
2. Press down arrow key, it should collapse to end of selection, but instead lands somewhere else.

**Comment by oyeanuj** [comment]
Here is another related  but simplified case to reproduce - https://jsfiddle.net/scef2uxx/2/

If you add it at the beginning of the block, the cursor lands at the wrong position. If you do it further in the sentence, the cursor is at the right place.

Also, seems like #630 has the same root issue.

**Comment by alexeygolev** [comment]
This is still relevant in Chrome 58.0.3029.96

**Comment by joshmarinacci** [comment]
I can reproduce this in Safari too

**Comment by joshmarinacci** [comment]
In fact, all of the  Transform.collapseTo variants fail. They all just jump the cursor to the start of the selection. Even if I don't call collapseTo at all the cursor moves after I insert a link.

...

Actually, I think the problem is related to focus/blur when we use Window.prompt. If I hard code the href and don't open a prompt window then it works like I expect.  Perhaps opening Window.prompt, at least on Safari for MacOSX, screws with the cursor location, so it resets to the start.

**Comment by nmerouze** [comment]
I found a workaround. I tried to create a link when the pasted text is a URL and the selection is collapsed. The [links](http://slatejs.org/#/links) example code would fail because `wrapInline` is deselecting the text so `collapseToEnd` has no effect. The selection needs to be selected again and extending to include all the pasted text again before calling `collapseToEnd`. I have made a gist: https://gist.github.com/nmerouze/7393ea75eb48a013a4c1d45ac07b6837

**Comment by nmerouze** [comment]
I found a workaround. I tried to create a link when the pasted text is a URL and the selection is collapsed. The links example code would fail because `wrapInline` is doing something to the selection which leads to `collapseToEnd` having no effect. The selection needs to be selected again and moved after the text before collapsing. I have made a gist: https://gist.github.com/nmerouze/7393ea75eb48a013a4c1d45ac07b6837
